### PR TITLE
exchange/ratelimiter: shift and standardise rate limiting functionality to requester.

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

Due to #1112 new shutdown functionality, it exposed a flaw in the exchange rate-limiting system. A `time.Sleep` would hang that routine and it wouldn't shutdown in a timely fashion and release resources.

This is opened as a draft so I don't forget; until #1112 is merged I can test these changes in their entirety. 

This is in review mode just to gauge feedback and whoopsie daisy design issues. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/request/limit.go [hunk 1]
```diff
@@ -74,11 +76,59 @@ func (r *Requester) InitiateRateLimit(ctx context.Context, e EndpointLimit) erro
 		return nil
 	}
 
-	if r.limiter != nil {
-		return r.limiter.Limit(ctx, e)
+	if r.limiter == nil {
+		return nil
```
Comment (gloriousCode) [inline exchanges/request/limit.go:80]: I think this should return err. Weird to call "initial rate limit" when its unset. Add protection surrounding the caller in `doRequest`

File: exchanges/request/limit.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ type EndpointLimit int
 // wrapper for extended rate limiting configuration i.e. Shells of rate
 // limits with a global rate for sub rates.
 type Limiter interface {
-	Limit(context.Context, EndpointLimit) error
+	Limit(context.Context, EndpointLimit) (*rate.Limiter, int, error)
```
Comment (gloriousCode) [inline exchanges/request/limit.go:45]: Signature could potentially even be a uint16, stronger typed and we can never have negative numbers assigned

**Review by gloriousCode — COMMENTED** [review]
I like this concept! Plus I like the one area to handle all limiting. This helps with our old bugs when we manage to get ratelimited and can't close the application in a peaceful manner

**Comment by shazbert** [comment]
Superseded by #1477

**Pull request closed without merging**